temba/channels/handlers.py

```diff
@@ -1115,24 +1116,58 @@ def post(self, request, *args, **kwargs):
 
         # this is a new incoming message
         elif action == 'receive':
-            if not any(attr in body for attr in ('timestamp', 'from_addr', 'content', 'message_id')):
-                return HttpResponse("Missing one of timestamp, from_addr, content or message_id, ignoring message", status=400)
+
+            if any(attr not in body for attr in ('timestamp', 'from_addr', 'message_id')):
+                return HttpResponse("Missing one of timestamp, from_addr or message_id, ignoring message",
+                                    status=400)
 
             # dates come in the format "2014-04-18 03:54:20.570618" GMT
             message_date = datetime.strptime(body['timestamp'], "%Y-%m-%d %H:%M:%S.%f")
             gmt_date = pytz.timezone('GMT').localize(message_date)
 
-            status = PENDING
-            if body.get('session_event') == "close":
-                status = INTERRUPTED
-            message = Msg.create_incoming(channel, URN.from_tel(body['from_addr']), body['content'], date=gmt_date, status=status)
+            content = body.get('content')
+
+            if is_ussd:  # receive USSD message
+                if body.get('session_event') == "close":
+                    status = USSDSession.INTERRUPTED
+                elif body.get('session_event') == "new":
+                    status = USSDSession.TRIGGERED
+                else:  # "resume" or null handling
+                    status = USSDSession.IN_PROGRESS
+
+                # determine the session ID
+                # since VUMI does not provide a session ID we have to fabricate it from some unique identifiers
+                # part1 - urn of the sender
+                # part2 - when the session was started or ordinal date
+                session_id_part1 = int(body.get('from_addr'))
```

Comment:
@smn  would it be possible to introduce a session id from your end? I assume if you have a session metadata you store the whole session with a unique id.